quilt3distribute/dataset.py

```diff
@@ -252,30 +307,99 @@ def distribute(
+                            meta = {}
+                            for index_col in self.index_columns:
+                                # Short reference to current metadata value
+                                v = v_ds.data[index_col].values[i]
+
+                                # Enforce simple JSON serializable type
+                                # First check if value is a numpy value
+                                # It likely is because pandas relies on numpy
+                                # All numpy types have the "dtype" attribute and can be cast to python type by using
+                                # the `item` function, details here:
+                                # https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.item.html
+                                if hasattr(v, "dtype"):
+                                    v = v.item()
+                                if isinstance(v, JSONSerializableTypes):
+                                    meta[index_col] = [v]
+                                else:
+                                    raise TypeError(
+                                        f"Non-simple-JSON-serializable type found in column: '{index_col}', "
+                                        f"at index: {i}: ({type(v)} '{v}').\n\n "
+                                        f"At this time only the following types are allowing in metadata: "
+                                        f"{JSONSerializableTypes}"
+                                    )
+
+                            # Check if object already exists
+                            if lk in pkg:
+                                # Join the two meta dictionaries
+                                joined_meta = {}
+                                for index_col, curr_v in pkg[lk].meta.items():
+                                    # Join the values for the current iteration of the metadata
+                                    joined_values = [*curr_v, *meta[index_col]]
+
+                                    # Only check if the metadata at this index can be reduced if currently is still
+                                    # being decided. We know if the metadata value at this index is still be decided if:
+                                    # the boolean value in the metadata reduction map is True, as in, this index can be
+                                    # reduced or collapsed.
+                                    # The other reason to make this check is so that we don't override an earlier False
+                                    # reduction value. In the case where early on we encounter an instance of the
+                                    # metadata that should not be reduced but then later on we say it can be, this check
+                                    # prevents that. As we want all metadata access across the dataset to be uniform.
+                                    if metadata_reduction_map[index_col]:
```

Comment:
Even with the description above, I'm still not clear what's going on here. All I can tell here is that we've encountered a duplicate file in the package and you've merged the metadata for that file with the prior instances of it in the same list.

Reply:
That is exactly it. That's all I have done. The only additional check is that we want to see if all the values in that merged list are still unique.

Reply:
I agree - this would be much clearer in a separate function with tests that might be associated directly to that piece of code to see what's expected out of this section.

Reply:
Ah, OK. Yes, I'm starting to understand. I agree with Matt--separating the functionality that checks for unique metadata values into its own named function would have helped me read this better. And...could just use a `set`?

Reply:
Will again come talk about why `set` cannot be used in this case.